File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/function/Cache.java
package com.jmsoftware.maf.springcloudstarter.function;

import java.util.Map;
import java.util.function.Function;

/**
 * <h1>Cache</h1>
 * <p>
 * Change description here.
 *
 * @author Johnny Miller (鍾俊), email: johnnysviva@outlook.com, 9/28/21 8:51 AM
 * @see <a href='https://juejin.cn/post/6892298625058078727#heading-2'>Java 函数式编程最佳实践 - 赋予函数缓存能力</a>
 **/
public class Cache {
    private Cache() {
    }

    /**
     * Cache function r.
     *
     * @param <T>      the type parameter
     * @param <R>      the type parameter
     * @param function the function
     * @param t        the t
     * @param cache    the cache
     * @return the r
     * @see com.jmsoftware.maf.springcloudstarter.FunctionalInterfaceTests#testCacheFunction()
     */
    public static <T, R> R cacheFunction(Function<T, R> function, T t, Map<T, R> cache) {
        R r = cache.get(t);
        if (r != null) {
            return r;
        }
        R result = function.apply(t);
        cache.put(t, result);
        return result;
    }

    /**
     * Cache function r.
     *
     * @param <T>      the type parameter
     * @param <R>      the type parameter
     * @param function the function
     * @param t        the t
     * @param cache    the cache
     * @return the r
     * @see com.jmsoftware.maf.springcloudstarter.FunctionalInterfaceTests#testCacheFunction()
     */
    public static <T, R> R cacheFunction(Function<T, R> function, T t, com.google.common.cache.Cache<T, R> cache) {
        R r = cache.getIfPresent(t);
        if (r != null) {
            return r;
        }
        R result = function.apply(t);
        cache.put(t, result);
        return result;
    }
}

File: common/src/main/kotlin/com/jmsoftware/maf/common/function/Cache.kt
@file:Suppress("unused")

package com.jmsoftware.maf.common.function

import com.google.common.cache.Cache
import java.util.function.Function

/**
 * # Cache
 *
 * Change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/14/22 2:44 PM
 * @see <a href='https://juejin.cn/post/6892298625058078727#heading-2'>Java 函数式编程最佳实践 - 赋予函数缓存能力</a>
 */
private object Cache

/**
 * Cache function r.
 *
 * @param <T>      the type parameter
 * @param <R>      the type parameter
 * @param function the function
 * @param t        the t
 * @param cache    the cache
 * @return the r
 */
fun <T, R> cacheFunction(function: Function<T, R>, t: T, cache: MutableMap<T, R>): R {
    val r = cache[t]
    if (r != null) {
        return r
    }
    val result = function.apply(t)
    cache[t] = result
    return result
}

/**
 * Cache function r.
 *
 * @param <T>      the type parameter
 * @param <R>      the type parameter
 * @param function the function
 * @param t        the t
 * @param cache    the cache
 * @return the r
 */
fun <T : Any, R : Any> cacheFunction(function: Function<T, R>, t: T, cache: Cache<T, R>): R {
    val r = cache.getIfPresent(t)
    if (r != null) {
        return r
    }
    val result = function.apply(t)
    cache.put(t, result)
    return result
}
